### PR TITLE
Use `cilium-envoy-starter` to run Envoy without privileges

### PR DESCRIPTION
**Pull request by jrajahalme**
Introduce a `cilium-envoy-starter` that forks and execs `cilium-envoy`
after having dropped all privileges.

Perform bpf map and privileged socket option operations in
`cilium-envoy-starter`, as requested by Cilium filters running in
`cilium-envoy` via a pipe between the two. Currently the protocol over
this pipe supports only the following operations:

- dump capabilities cilium-envoy-starter is running with
- open bpf maps
- perform bpf map lookups
- set socket options

cilium-envoy-starter fails to start if it does not have adequate
privileges needed for the above operations.

cilium-envoy now exits if it is running with any privileges when Cilium
filters are first configured.

Implementation detail:

- `libcap` is not used, as it would be hard to cross-compile for both arm64
  and amd64. Fortunately we only need to drop capabilities, so using the
  system call interface is not complicated.

- Patch Envoy to support setting socket options from listener filters and
  use them to set privileged options for the listener.

- Remove the unused support for creating and manipulating bpf maps.

- Remove `setrlimit` call that is only needed if the process is creating
  bpf maps.

- Minor cleanup in socket option implementation to move unused fields
  from SocketMarkOption class to SocketOption where they are used.

### CODE REVIEW

**Comment by jrajahalme** [comment]
Cilium Integration test fails expectedly as Cilium Agent update is also needed so that it will launch Envoy via the new `cilium-envoy-wrapper`. Could also consider reversing the names so that `cilium-envoy` would be the wrapper and actual Envoy binary would be e.g., `cilium-envoy-wrapped`, but IMO it is clearer the way it is now in this PR.

**Comment by jrajahalme** [comment]
> Cilium Integration test fails expectedly as Cilium Agent update is also needed so that it will launch Envoy via the new cilium-envoy-wrapper. Could also consider reversing the names so that cilium-envoy would be the wrapper and actual Envoy binary would be e.g., cilium-envoy-wrapped, but IMO it is clearer the way it is now in this PR.

Cilium Agent update is anyway needed to drop the socket options and `Transparent=true` from Listener config, so there is no way we can do this without agent update. Might as well go with the cleaner executable names.

File: cilium/envoy_wrapper_main.cc [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+} // namespace CiliumEnvoyWrapper
+
+// NOLINT(namespace-envoy)
+
+#define WRAPPER_SUFFIX "-wrapper"
+#define WRAPPER_SUFFIX_LEN (sizeof(WRAPPER_SUFFIX) - 1)
+
+int main(int /* argc */, char** argv) {
```
Comment (mhofstetter) [inline cilium/envoy_wrapper_main.cc:171]: after having a first look at the overall structure (new bazel libs and the `main` of the wrapper) i wonder whether it wouldn't be better to even further separate the wrapper logic from the "privileged service" bits.

Currently:
```mermaid
graph LR;
    W[Wrapper]-- forks -->E[Envoy];
    E-. calls .-> W;
```

Idea:
```mermaid
graph LR;
    W[Wrapper]-- forks -->E[Envoy];
    W[Wrapper]-- forks -->P[Privileged Service];
    E-. calls .-> P;
```

IMO this could lead to better understandable architecture and names (bazel libraries, C++ types, etc). Currently everything is named ~ `EnvoyWrapper**` which is kind of unclear.

* Wrapper: Being responsible to start the necessary processes. Maybe we could even rename this to "EnvoyStarter" to make this more clear.
* Privileged Service: Responsible to provide the privileged actions (BPF map access, socket handling) to the Envoy process. (Same here, we might find a more precise name)

And it might even support cases where the two processes (envoy & "privileged service") are started independent of the wrapper (starter) - e.g. by running the privileged service in a separate sidecar. Or at least we're preparing for this from an architectural point of view.

**Review by mhofstetter — COMMENTED** [review]
Nice - already looks great! :rocket: 

I placed an "architectural" question (with a proposal) about the wrapper inline. Might be worth to discuss this first before I dive deeper into the details.

File: cilium/envoy_wrapper_main.cc (hunk 1 above)
Comment (jrajahalme) [reply cilium/envoy_wrapper_main.cc:171]: > And it might even support cases where the two processes (envoy & "privileged service") are started independent of the wrapper (starter) - e.g. by running the privileged service in a separate sidecar. Or at least we're preparing for this from an architectural point of view.

This will need a different, more explicit method of passing the fd to the cilium-envoy process. I looked into this briefly, but did not test anything out yet, IMO we can leave this for now. However, if we implement support for the sidecar then the wrapper/starter is not necessary at all. So IMO building more structure into this now would be an overkill.

Better names for the bazel bits are welcomed, I'm pretty bad at naming these kind of things :-)

**Comment by jrajahalme** [comment]
> Nice - already looks great! 🚀
> 
> I placed an "architectural" question (with a proposal) about the wrapper inline. Might be worth to discuss this first before I dive deeper into the details.

Detail would not change much if the wrapper would fork another process to serve the privileged requests, basically one more fork and the parent process waiting for this new process to finish - I think this is unnecessary, however, and would not help architecturally either.

So please delve into the detail :-)

File: cilium/envoy_wrapper_main.cc (hunk 1 above)
Comment (mhofstetter) [reply cilium/envoy_wrapper_main.cc:171]: > > And it might even support cases where the two processes (envoy & "privileged service") are started independent of the wrapper (starter) - e.g. by running the privileged service in a separate sidecar. Or at least we're preparing for this from an architectural point of view.
> 
> This will need a different, more explicit method of passing the fd to the cilium-envoy process. I looked into this briefly, but did not test anything out yet, IMO we can leave this for now. However, if we implement support for the sidecar then the wrapper/starter is not necessary at all. So IMO building more structure into this now would be an overkill.
> 
> Better names for the bazel bits are welcomed, I'm pretty bad at naming these kind of things :-)

I agree that the sidecar idea is a bit off for the time being. This shouldn't be the reason.

But as mentioned - it would also help in better understanding the design by separating the responsibilities - and therefore allow for better naming. Naming is especially hard if "the thing" is responsible for multiple tasks. But we might can solve this by splitting up `envoy_wrapper_main.cc` into these two aspects.

Comment (mhofstetter) [reply cilium/envoy_wrapper_main.cc:171]: I'd recommend to separate the main function (which is the actual wrapper / starter) from the "ProtocolServer by moving the ProtocolServer bits into its own file `envoy_wrapper_protocolserver.cc`.

And renaming all `envoy_wrapper_*`  (except `envoy_wrapper_main.cc`)  files to something `privileged_service_*` 

-> 

* `privileged_service_client`
* `privileged_service_server`
* `privileged_service_protocol`

and adapt the bazel envoy libs (and their names accordingly)

File: cilium/envoy_wrapper_main.cc [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+  if (pid == 0) {
+    // in child process, close the parent end of the pipe
+    close(fds[0]);
+
+    // Unconditionally drop all capabilities
+    struct __user_cap_header_struct hdr{_LINUX_CAPABILITY_VERSION_3, 0};
+    struct __user_cap_data_struct data[2];
+    memset(&data, 0, sizeof(data));
+    if (::syscall(SYS_capset, &hdr, &data, sizeof(data)) != 0) {
+      perror("capset");
```
Comment (mhofstetter) [inline cilium/envoy_wrapper_main.cc:219]: Provide more meaningful error messages in all the `perror` calls

File: cilium/envoy_wrapper_main.cc (hunk 1 above)
Comment (jrajahalme) [reply cilium/envoy_wrapper_main.cc]: Renamed and separated as suggested. Also renamed the new binary as `cilium-envoy-starter`, but kept it in a single process that now implements the `PrivilegedServiceServer`.

File: cilium/envoy_wrapper_main.cc (hunk 2 above)
Comment (mhofstetter) [reply cilium/envoy_wrapper_main.cc]: sounds reasonable!

File: starter/privileged_service_client.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+public:
+  ProtocolClient();
+
+  // allow access to the classes that need it
+  friend class Envoy::Cilium::Bpf;
+  friend class Envoy::Cilium::SocketMarkOption;
+
+protected:
+  // Read-only bpf syscalls
+  Envoy::Api::SysCallIntResult bpf_open(const char *path);
```
Comment (mhofstetter) [inline starter/privileged_service_client.h:31]: document the 3 "public" facing functions?

File: starter/privileged_service_client.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  Envoy::Api::SysCallIntResult bpf_lookup(int fd, const void *key, uint32_t key_size, void* value,
+					  uint32_t value_size);
+
+  // Set a socket option
+  Envoy::Api::SysCallIntResult setsockopt(int sockfd, int level, int optname, const void *optval,
+					  socklen_t optlen);
+
+private:
+  bool check_privileged_service();
+  bool running_with_cilium_privileged_service() const { return is_open(); }
```
Comment (mhofstetter) [inline starter/privileged_service_client.h:41]: rename to `connected_to_cilium_privileged_service` ?

File: starter/privileged_service_server.cc [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+		     msg.setsockopt_req.optname_, &msg.setsockopt_req.optval_,
+		     sizeof(msg.setsockopt_req.optval_));
+      break;
+    default:
+      fprintf(stderr, "Unexpected privileged call type: %d\n", msg.hdr.msg_type_);
+      rc = -1;
+      errno = EINVAL;
+    }
+
+    // Close the received file descriptor
```
Comment (mhofstetter) [inline starter/privileged_service_server.cc:118]: nit: `receiving file descriptor`?

File: starter/privileged_service_server.cc [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      msg.response.ret_ = {INT_MAX, 0};
+    } else {
+      msg.response.ret_ = {rc, rc != -1 ? 0 : errno};
+    }
+    size = send_fd_msg(&msg, sizeof(msg.response), buf, value_len, fd_out);
+    if (size < ssize_t(sizeof(msg.response) + value_len)) {
+      perror("sendmsg");
+    }
+
+    // Close the sent file descriptor
```
Comment (mhofstetter) [inline starter/privileged_service_server.cc:137]: nit: `sending file descriptor`?

File: cilium/envoy_wrapper_main.cc (hunk 1 above)
Comment (mhofstetter) [reply cilium/envoy_wrapper_main.cc]: nice - looks great! thanks!

**Review by mhofstetter — CHANGES_REQUESTED** [review]
Nice improvement by splitting out the privileged service (&starter)! :rocket: 

Some suggestions from the next review iteration inline.

File: cilium/bpf.cc [hunk 7]
```diff
@@ -39,10 +25,10 @@ void Bpf::close() {
 
 bool Bpf::open(const std::string& path) {
   bool log_on_error = ENVOY_LOG_CHECK_LEVEL(trace);
-  union bpf_attr attr = {};
-  attr.pathname = uintptr_t(path.c_str());
 
-  fd_ = bpfSyscall(BPF_OBJ_GET, &attr);
+  auto& cilium_calls = PrivilegedService::Singleton::get();
```
[the remark this answers is not in the document]
Comment (jrajahalme) [reply cilium/bpf.cc:29]: Not sure, I kind of like to keep this name Cilium specific (see other comments below).

[the remark this answers is not in the document]
Comment (mhofstetter) [reply cilium/bpf.cc:29]: `cilium_privileged_service`?

File: starter/privileged_service_client.h (hunk 3 above)
Comment (jrajahalme) [reply starter/privileged_service_client.h:31]: I'm thinking this privileged API will morph to something more CIlium specific in near future. Giving access to set any/all socket options and open any bpf/all bpf maps is not the most secure thing to do. Both the bpf lookup and socket option setting happen via the bpf_metadata filter, so there would be an opportunity to lift the abstraction level of this privileged API from the current syscall level to something like "Cilium connection metadata setup" that would:
- Get the metadata for the new connection from conntrack and/or ipcache maps, and, if successful
- Set the socket options Cilium needs for new proxied connections
A separate call would be needed to set up the listening socket (socket options only)

This is a bit bigger change, shifting all the bpf code to the starter and refactoring the bpf_metadata filter code, so I did not want to start with this "bigger picture".

So in summary, I'd like to skip documenting this version of the privileged API, and would like to not characterise this as "public facing", even though it kind of is.

File: starter/main.cc [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    if (fds[1] != CILIUM_PRIVILEGED_SERVICE_FD) {
+      if (dup2(fds[1], CILIUM_PRIVILEGED_SERVICE_FD) < 0) {
+	perror("dup2");
+	exit(1);
+      }
+      close(fds[1]);
+    }
+    
+    // Exec cilium-envoy process
+    execv(path, argv);
```
[the remark this answers is not in the document]
Comment (jrajahalme) [reply starter/main.cc:89]: I've tried. Currently the server side exits if the client side closes the socket, and the destructor you mention waits for the client process to exit. Client side closes the socket on destruction, but maybe this never happens as Envoy:: ThreadSafeSingleton seems to not ever destruct.

However, the client side of the socket will get closed when the process exits, so the starter should gracefully exit right after, and the wait for the client process should be a no-op as it has already exited.

File: starter/privileged_service_server.cc (hunk 5 above)
Comment (jrajahalme) [reply starter/privileged_service_server.cc:118]: No, it is literally closing the file descriptor received on the control channel of `recvmsg`.

File: starter/privileged_service_server.cc (hunk 6 above)
Comment (jrajahalme) [reply starter/privileged_service_server.cc:137]: Same here, when opening a bpf map, the opened file descriptor is sent in recvmsg control channel and we need to close the sent file descriptor, as otherwise the bpf map would stay open due to the server side keeping an open file descriptor to it.

File: starter/main.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (mhofstetter) [reply starter/main.cc:89]: great, thanks for the explanation!

File: starter/privileged_service_client.h (hunk 3 above)
Comment (mhofstetter) [reply starter/privileged_service_client.h:31]: makes sense. thanks for the outlook and explanation!

File: starter/privileged_service_client.h (hunk 4 above)
Comment (jrajahalme) [reply starter/privileged_service_client.h:41]: IMO this was (already) getting too long, renamed as `have_cilium_privileged_service()`

File: starter/privileged_service_client.cc [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+  DumpRequest req;
+  Response resp;
+  uint8_t buf[1024];
+  int fd = -1;
+  ssize_t size = transact(req.hdr_, sizeof(req), nullptr, 0, &fd, resp, buf, sizeof(buf), false);
+  if (size < ssize_t(sizeof(resp))) {
+    return false;
+  }
+  std::string str(reinterpret_cast<char *>(buf), size - sizeof(resp));
+  ENVOY_LOG_MISC(debug, "Running with Cilium privileged service with the following capabilities: {}", str);
```
[the remark this answers is not in the document]
Comment (jrajahalme) [reply starter/privileged_service_client.cc:77]: rephrased as: `Cilium privileged service detected with following capabilities: {}`

**Review by mhofstetter — APPROVED** [review]
LGTM :rocket:

**Comment by jrajahalme** [comment]
Cilium connectivity tests fail expectedly due to Envoy having been started with privileges without `cilium-envoy-starter`:
```
2023-09-11T17:18:27.392970338Z level=error msg="[assert failure: get_capabilities(CAP_EFFECTIVE) == 0 && get_capabilities(CAP_PERMITTED) == 0. Details: cilium-envoy running with privileges, exiting" subsys=envoy-assert threadID=979
```
Merging to main and posting an update on `cilium/cilium` to pick this up :-)
